services/replicator/metadataReconciler.go

```diff
@@ -451,37 +470,170 @@ func (r *metadataReconciler) reconcileDestExtent(destUUID string, localExtents m
-				r.m3Client.UpdateGauge(metrics.ReplicatorReconcileScope, metrics.ReplicatorReconcileDestExtentInconsistentStatus, 1)
+			if remoteExtentStatus == shared.ExtentStatus_SEALED && localExtentStatus == shared.ExtentStatus_OPEN {
+				r.sealExtentInMetadata(destUUID, remoteExtentUUID)
+			}
+			if remoteExtentStatus == shared.ExtentStatus_DELETED {
+				r.handleExtentDeletedOrMissingInRemote(destUUID, remoteExtentUUID, localExtentStatus)
+			}
+		}
+	}
 
-				updateRequest := &metadata.UpdateExtentStatsRequest{
-					ExtentUUID:      common.StringPtr(remoteExtentUUID),
-					DestinationUUID: common.StringPtr(destUUID),
-					Status:          common.MetadataExtentStatusPtr(shared.ExtentStatus_SEALED),
-				}
-				_, err := r.mClient.UpdateExtentStats(nil, updateRequest)
-				if err != nil {
-					r.logger.WithFields(bark.Fields{
-						common.TagErr: err,
-						common.TagDst: common.FmtDst(destUUID),
-						common.TagExt: common.FmtExt(remoteExtentUUID),
-					}).Error(`Failed to update extent status to sealed`)
-					continue
-				}
+	// now try to find all the orphaned extents(extent already gone in remote zone but not in local zone), and try to seal them
 
-				r.logger.WithFields(bark.Fields{
-					common.TagDst: common.FmtDst(destUUID),
-					common.TagExt: common.FmtExt(remoteExtentUUID),
-				}).Info(`Extent SEALED`)
+	// note we're not 100% sure if the extent is really gone even if cassandra doesn't return that extent because of
+	// the eventual consistency nature of cassandra
+	// so we maintain a timestamp to track how long the extent has been missing
+	// we act on it only after it has been missing for a certain period of time
+
+	remoteMissingExtents := make(map[string]bool)
+	for localExtentUUID, localExtentStatus := range localExtents {
+		// we're going to delete this extent soon locally so no need to act on it
+		if localExtentStatus == shared.ExtentStatus_CONSUMED || localExtentStatus == shared.ExtentStatus_DELETED {
+			continue
+		}
+		if _, ok := remoteExtents[localExtentUUID]; !ok {
+			// extent exists in local but not in remote
+			remoteMissingExtents[localExtentUUID] = true
+		}
+	}
+
+	for suspectExtent, suspectExtentInfo := range r.suspectMissingExtents {
+		if suspectExtentInfo.destUUID != destUUID {
+			continue
+		}
+		if _, ok := remoteMissingExtents[suspectExtent]; !ok {
+			// re-appeared in remote, remove it from suspect list
+			delete(r.suspectMissingExtents, suspectExtent)
+		} else {
+			if time.Since(suspectExtentInfo.missingSince) > extentMissingDurationThreshold {
+				r.handleExtentDeletedOrMissingInRemote(suspectExtentInfo.destUUID, suspectExtent, localExtents[suspectExtent])
+			}
+		}
+	}
 
+	// now add new suspect if there's any
+	for missingExtent, _ := range remoteMissingExtents {
+		if _, ok := r.suspectMissingExtents[missingExtent]; !ok {
+			r.suspectMissingExtents[missingExtent] = missingExtentInfo{
+				destUUID:     destUUID,
+				missingSince: time.Now(),
 			}
 		}
 	}
+
```

Comment:
also, would be cool to emit a metric on len(suspectMissingExtents) at the end of each reconciliation.